### PR TITLE
📦 NEW: Upgrade webpack and @babel/core

### PR DESCRIPTION
**Pull request by KenEucker**
This is to address an issue I created: ahmadawais/create-guten-block#149 

This is primarily a dependency update but also includes code changes to the eject.js file for the ejected package.json.

This code includes the similar changes made in @asharirfan's fork(https://github.com/ahmadawais/create-guten-block/commits/asharirfan/webpack-4).

One thing I did not see that I expected to make changes to is the babel configuration used in the build and start tasks. If I missed that and need to add the updated plugins and presets to that as well, let me know.

EDIT: I've tested this with my own project using multiple blocks.

Thanks, team! 💖

### CODE REVIEW

**Comment by ahmadawais** [comment]
Thank you. 

Also mention how you tested this PR. Default blocks work? Custom one's work?

File: packages/cgb-scripts/scripts/eject.js [hunk 1]
```diff
@@ -273,27 +274,25 @@ inquirer
 				],
 			],
 			plugins: [
-				[ 'transform-object-rest-spread' ],
+				'@babel/plugin-syntax-dynamic-import',
+				'@babel/plugin-syntax-import-meta',
 				[
-					'transform-object-rest-spread',
+					'@babel/plugin-proposal-class-properties',
```
Comment (ahmadawais) [inline packages/cgb-scripts/scripts/eject.js]: Why did you remove

- transform-object-rest-spread
- transform-object-rest-spread
- transform-react-jsx
- transform-runtime

🤔

File: packages/cgb-scripts/package.json [hunk 2]
```diff
@@ -45,7 +56,7 @@
 		"eslint-plugin-jsx-a11y": "^6.0.3",
 		"eslint-plugin-react": "^7.5.1",
 		"eslint-plugin-wordpress": "^0.1.0",
-		"extract-text-webpack-plugin": "^3.0.2",
+		"extract-text-webpack-plugin": "^4.0.0-beta.0",
```
Comment (ahmadawais) [inline packages/cgb-scripts/package.json:61]: Does this work in the same way? I mean is it backward compatible for so many plugins built on top of create-guten-block?

File: packages/cgb-scripts/package.json [hunk 3]
```diff
@@ -30,10 +30,21 @@
...
+		"@babel/plugin-syntax-import-meta": "^7.0.0",
+		"@babel/preset-env": "^7.4.4",
+		"@babel/preset-react": "^7.0.0",
 		"autoprefixer": "^7.2.4",
-		"babel-core": "^6.25.0",
-		"babel-eslint": "^8.2.1",
-		"babel-loader": "^7.1.1",
+		"babel-eslint": "^9.0.0",
+		"babel-loader": "^8.0.0",
 		"babel-preset-cgb": "^1.6.0",
```
Comment (ahmadawais) [inline packages/cgb-scripts/package.json:50]: I believe we might need to update `babel-preset-cgb` as well.

Comment (ahmadawais) [reply packages/cgb-scripts/package.json:50]: Check this https://github.com/ahmadawais/create-guten-block/blob/master/packages/babel-preset-cgb/index.js

**Comment by ahmadawais** [comment]
For the sake of testing I just published.

```
Successfully published:
 - babel-preset-cgb@1.7.0-alpha.f64d360b
 - cgb-dev-utils@1.6.0-alpha.f64d360b
 - cgb-scripts@1.18.0-alpha.f64d360b
 - create-guten-block@1.13.0-alpha.f64d360b
```

You can now run

```sh
npx create-guten-block@1.13.0-alpha.f64d360b myBlock --canary

# OR

npx create-guten-block@1.13.0-alpha.f64d360b myBlock -c
```

and test this PR.

**Review by ahmadawais — CHANGES_REQUESTED** [review]
I just tested this PR and it doesn't compile. 

Here's the error


```sh

❌   Failed to compile.


👉  ./src/blocks.js
Module build failed (from ./node_modules/babel-loader/lib/index.js):
Error: Plugin/Preset files are not allowed to export objects, only functions. In /Users/ahmadawais/Documents/Web/Dev/Local/local/app/public/wp-content/plugins/myblock/node_modules/babel-preset-cgb/index.js
    at Array.map (<anonymous>)
    at Generator.next (<anonymous>)


⠼ Watching for changes... let's fix this... (Press CTRL + C to stop).


```

img

**Review by ahmadawais — CHANGES_REQUESTED** [review]
Moreover, I see that this is outdated

```sh
❯◯ babel-eslint        9.0.0    ❯  10.0.1  https://github.com/babel/babel-eslint
```

File: packages/cgb-scripts/scripts/eject.js (hunk 1 above)
Comment (KenEucker) [reply packages/cgb-scripts/scripts/eject.js]: When running babel-upgrade, I expected the previous plugins to be included but it's possible that I am mistaken about that. 

I know that transform-react-jsx is included with 'preset-react' and I simply assumed that transform-runtime and transform-object-spread were similarly included in other ways:
https://babeljs.io/docs/en/babel-preset-react

In running this against my own guten blocks, everything compiled and is working correctly though it seems that more in depth testing may be needed.

**Comment by KenEucker** [comment]
RE: "Error: Plugin/Preset files are not allowed to export objects, only functions."

It looks like babel-preset-cgb may need to be updated to work with Babel 7? Am I reading that incorrectly?

**Comment by ahmadawais** [comment]
Yes that's right.

File: packages/cgb-scripts/package.json (hunk 2 above)
Comment (KenEucker) [reply packages/cgb-scripts/package.json:61]: This one was hard to determine as that package appears to be outdated and unmaintained. However, if I remember correctly, v4.0.0-beta.0 is the only one that works with webpack 4.

Ultimately, I believe that removing the extract-text-webpack-plugin dependency and replacing it with mini-css-extract plugin is the better route to go.

File: packages/cgb-scripts/package.json (hunk 3 above)
Comment (KenEucker) [reply packages/cgb-scripts/package.json:50]: ^ yes that's exactly what I was looking for that I was missing. I understand now how the preset is used for the build task before ejection, and then not used at all afterwards. It seems like it might be a better solution to grab the babel-preset-cgb and use that inside eject.js instead of duping this code in both places. 

Working on this change now.

**Comment by KenEucker** [comment]
@ahmadawais I've addressed the changes you requested. Please let me know if I've missed anything or if there's more you're looking for. Thanks!

File: packages/babel-preset-cgb/index.js [hunk 4]
```diff
@@ -30,35 +29,38 @@ module.exports = {
...
 				useBuiltIns: true,
 			},
 		],
 		[
-			// Transforms JSX Syntax.
-			// 'transform-react-jsx',
-			require.resolve( 'babel-plugin-transform-react-jsx' ),
+			'@babel/plugin-proposal-class-properties',
 			{
-				pragma: 'wp.element.createElement',
```
Comment (ahmadawais) [inline packages/babel-preset-cgb/index.js:49]: Why removing the pragma? It won't work without this.

File: packages/babel-preset-cgb/index.js [hunk 5]
```diff
@@ -30,35 +29,38 @@ module.exports = {
...
+				loose: true,
 			},
 		],
-		// Async/Await awesomeness https://babeljs.io/docs/en/babel-plugin-syntax-async-functions/.
-		[ require.resolve( 'babel-plugin-syntax-async-functions' ) ],
```
Comment (ahmadawais) [inline packages/babel-preset-cgb/index.js:53]: Why removing this?

File: packages/cgb-scripts/package.json [hunk 6]
```diff
@@ -55,10 +68,10 @@
...
 		"raw-loader": "^0.5.1",
 		"resolve-pkg": "^1.0.0",
-		"sass-loader": "^6.0.6",
-		"shelljs": "^0.8.0",
+		"sass-loader": "^7.1.0",
+		"shelljs": "^0.8.3",
 		"style-loader": "^0.19.1",
 		"update-notifier": "^2.3.0",
-		"webpack": "^3.1.0"
+		"webpack": "^4.30.0"
```
Comment (ahmadawais) [inline packages/cgb-scripts/package.json:75]: webpack 4 means that the current config will not run since the CSS Extraction is not supported the same way.

**Comment by ahmadawais** [comment]
@KenEucker Left some comments.

Please request a review after you test that all of this is working:

- The block plugin gets created without errors
- `npm start` runs and breaks if you break syntax in the block
- You can include the block in a post, change both back/frontend to check it works
- Check if css files of editor and frontend work by changing colors
- Finally, check if `npm run build` works

Looking forward, peace! ✌️

File: packages/babel-preset-cgb/index.js (hunk 5 above)
Comment (KenEucker) [reply packages/babel-preset-cgb/index.js:53]: In looking into this one, I discovered that there is no @babel equivalent of this plugin. Moreover, I believe the functionality desired is present, unless we need more than just supporting async functions. Do we need to include @babel/plugin-transform-async-to-generator ? 

------------ reference -----------------
https://github.com/babel/babel/issues/9017

"That package has been deprecated and is not maintained anymore, since parsing async functions is now enabled by default. If you want to transform them, you should use @babel/plugin-transform-async-to-generator."

File: packages/babel-preset-cgb/index.js (hunk 4 above)
Comment (KenEucker) [reply packages/babel-preset-cgb/index.js:49]: I've now added a preset that covers this(@wordpress/babel-preset-default), instead of the pragma option on a plugin we are no longer including (the @babel/react handles this).

File: packages/babel-preset-cgb/index.js (hunk 5 above)
Comment (ahmadawais) [reply packages/babel-preset-cgb/index.js:53]: since parsing async functions is now enabled by default we can just let go of it.

File: packages/cgb-scripts/package.json (hunk 2 above)
Comment (ahmadawais) [reply packages/cgb-scripts/package.json:61]: I am open to replace it, if you'd like to take charge here. 

Two options:

1. https://quantizd.com/webpack-4-extract-css-with-mini-css-extract-plugin/
2. https://stackoverflow.com/questions/50976979/output-2-or-more-css-files-with-mini-css-extract-plugin-in-webpack

I'm siding with the second option if the first one doesn't work. Go ahead and try that?

File: packages/cgb-scripts/scripts/eject.js (hunk 1 above)
Comment (ahmadawais) [reply packages/cgb-scripts/scripts/eject.js]: Have you test `-object-rest-spread` and it's working?

File: packages/cgb-scripts/package.json (hunk 3 above)
Comment (ahmadawais) [reply packages/cgb-scripts/package.json:50]: Go right ahead!

File: packages/babel-preset-cgb/index.js (hunk 4 above)
Comment (ahmadawais) [reply packages/babel-preset-cgb/index.js:49]: I don't understand this. Do we not need it anymore? Why? Since WP does use a pragma option here. I am happy to drop it if there's no need.

**Comment by ahmadawais** [comment]
@KenEucker feels like we're getting closer to pushing an update. Waiting on you to test out the stuff removed from Babel still works, and porting both dev/build webpack configs.

Once done, and tested — please request a review so I get an email.

Thank you! 👌

File: packages/babel-preset-cgb/index.js (hunk 4 above)
Comment (KenEucker) [reply packages/babel-preset-cgb/index.js:49]: @ahmadawais Digging even deeper I found this:

https://github.com/WordPress/gutenberg/blob/master/packages/babel-preset-default/index.js

Which appears to include what we need here. It includes the pragma settings. I'm beginning to test this now. It's weird that in my testing I was able to create blocks and insert them into posts without this part of the config. I'm going to spend a bit more time testing around this.

File: packages/cgb-scripts/package.json (hunk 2 above)
Comment (KenEucker) [reply packages/cgb-scripts/package.json:61]: @ahmadawais it looks like @asharirfan's PR addresses this and includes the mini-css-extract-plugin work:

https://github.com/ahmadawais/create-guten-block/pull/166

Should I merge that work into my branch and continue testing from there?

Comment (ahmadawais) [reply packages/cgb-scripts/package.json:61]: #166 is draft PR and WIP. You can start your own separate one, I test #166 and it was not working.

**Comment by KenEucker** [comment]
Checking back in on this to say that I've grown busier with work stuff in the last week. I am hoping to get more acquainted with testing this project and be able to have the additional changes queued up by next week. Thank you for your patience!
